### PR TITLE
Add toDecimalString function

### LINKED ISSUE #45249: formatDecimal function
Takes 2 argument - first is numeric, second is precision and returns a string

```
SELECT formatDecimal(2,2); -- 2.00
SELECT formatDecimal(2.123456,2); -- 2.12
SELECT formatDecimal(2.1456,2); -- 2.15 -- rounding!
SELECT formatDecimal(64.32::Float64, 2); -- 64.32
SELECT formatDecimal(64.32::Decimal64(2), 2); -- 64.32
SELECT formatDecimal(64.32::Decimal64(2), 2); -- 64.32
``` 

It should behave similar to [%f in sprintf](https://linux.die.net/man/3/sprintf), see also 

See also:
https://github.com/ClickHouse/ClickHouse/pull/27680#issuecomment-900762221
https://github.com/ClickHouse/ClickHouse/issues/30934#issuecomment-955998935

### PR DESCRIPTION
**Pull request by zvonand**
Added `toDecimalString` function allowing to convert numbers to string with fixed precision.

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added `toDecimalString` function allowing to convert numbers to string with fixed precision.

Closes #45249

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
Fuzzer has found errors.

File: src/Functions/FunctionFormatDecimal.h [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+        throw Exception(ErrorCodes::ILLEGAL_COLUMN, "Illegal column {} of argument of function {}",
+                        arguments[0].column->getName(), getName());
+    }
+
+private:
+    template <typename T, std::enable_if_t<is_integer<T> || is_floating_point<T>, bool> = true>
+    ColumnPtr executeType(const ColumnsWithTypeAndName & arguments) const
+    {
+        auto result_column_string = ColumnString::create();
+        auto col_to = assert_cast<ColumnString *>(result_column_string.get());
```
Comment (rschu1ze) [inline src/Functions/FunctionFormatDecimal.h:238]: Local clang-tidy complains this should be `auto * col_to`.

File: src/Functions/FunctionFormatDecimal.h [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+        ColumnString::Chars & data_to = col_to->getChars();
+        ColumnString::Offsets & offsets_to = col_to->getOffsets();
+
+        const auto * from_col = checkAndGetColumn<ColumnVector<T>>(arguments[0].column.get());
```
Comment (rschu1ze) [inline src/Functions/FunctionFormatDecimal.h:242]: `ColumnVector<T>` isn't exactly wrong. More generic would be `DataTypeNumber::ColumnType<T>`. And in l. 277: `DataTypeDecimal::ColumnType<T>`.

File: src/Functions/FunctionFormatDecimal.h [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+
+    static void format(double value, DB::WriteBuffer & out, int precision)
+    {
+        DB::DoubleConverter<false>::BufferType buffer;
+        double_conversion::StringBuilder builder{buffer, sizeof(buffer)};
+
+        const auto result = DB::DoubleConverter<false>::instance().ToFixed(value, precision, &builder);
+
+        if (!result)
+            throw DB::Exception(DB::ErrorCodes::CANNOT_PRINT_FLOAT_OR_DOUBLE_NUMBER, "Cannot print float or double number");
```
Comment (rschu1ze) [inline src/Functions/FunctionFormatDecimal.h:147]: Guess the exception msg could include `value` and `precision`.

File: src/Functions/FunctionFormatDecimal.h [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+        {
+            format(vec_from[i], buf_to, vec_precision[i]);
+            writeChar(0, buf_to);
+            offsets_to[i] = buf_to.count();
+        }
+
+        buf_to.finalize();
+    }
+
+    template <typename FirstArgType>
```
Comment (rschu1ze) [inline src/Functions/FunctionFormatDecimal.h:66]: `constantVector` is a weird use case ... not sure if it's needed.

File: src/Functions/FunctionFormatDecimal.h [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+    void constantVector(const FirstArgType & value_from, const ColumnVector<UInt8>::Container & vec_precision,
+                        ColumnString::Chars & vec_to, ColumnString::Offsets & offsets_to, const UInt8 from_scale) const
+    {
+        WriteBufferFromVector<ColumnString::Chars> buf_to(vec_to);
+        size_t input_rows_count = vec_precision.size();
+        offsets_to.resize(input_rows_count);
+
+        for (size_t i = 0; i < input_rows_count; ++i)
+        {
+            writeText(value_from, from_scale, buf_to, true, true, static_cast<UInt32>(vec_precision[i]));
```
Comment (rschu1ze) [inline src/Functions/FunctionFormatDecimal.h:130]: cast UInt8 --> UInt32 looks unnecessary

File: tests/queries/0_stateless/02676_format_decimal.sql [hunk 6]
```diff
@@ -0,0 +1,20 @@
+-- Regular types
+SELECT formatDecimal(2, 2);  -- more digits required than exist
+SELECT formatDecimal(2.123456, 2);  -- rounding
```
Comment (rschu1ze) [inline tests/queries/0_stateless/02676_format_decimal.sql:3]: Actually, the docs don't say that the fractional digits are potentially rounded (it sounds like they are truncated). Clarify that?

File: tests/queries/0_stateless/02676_format_decimal.sql [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+-- Decimals
+SELECT formatDecimal(32.345::Decimal32(3), 3);
+SELECT formatDecimal(32.345::Decimal32(3), 6);  -- more digits required than exist
+SELECT formatDecimal(32.456::Decimal32(3), 2);  -- rounding
+SELECT formatDecimal('64.5671232345'::Decimal64(10), 10);
+SELECT formatDecimal('128.78932312332132985464'::Decimal128(20), 20);
+SELECT formatDecimal('128.78932312332132985464123123'::Decimal128(26), 20);  -- rounding
+SELECT formatDecimal('128.78932312332132985464'::Decimal128(20), 26);  -- more digits required than exist
+SELECT formatDecimal('128.789323123321329854641231237893231233213298546'::Decimal256(45), 10);  -- rounding
+SELECT formatDecimal('128.789323123321329854641231237893231233213298546'::Decimal256(45), 50);  -- more digits required than exist
```
Comment (rschu1ze) [inline tests/queries/0_stateless/02676_format_decimal.sql:20]: Would be great to include a few negative tests (to check for exception handling)

File: src/Functions/FunctionFormatDecimal.h [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+        else
+        {
+            throw Exception(ErrorCodes::ILLEGAL_COLUMN, "Illegal column {} of first argument of function formatDecimal",
+                            arguments[0].column->getName());
+        }
+
+        return result_column_string;
+    }
+
+    template <typename T, std::enable_if_t<is_decimal<T>, bool> = true>
```
[the remark this answers is not in the document]
Comment (rschu1ze) [reply src/Functions/FunctionFormatDecimal.h:269]: Maybe class `ColumnVectorOrDecimalT` (Columns/ColumnDecimal.h) can help?

File: docs/en/sql-reference/functions/type-conversion-functions.md [hunk 9]
```diff
@@ -737,6 +737,41 @@ Result:
 └────────────┴───────┘
 ```
 
+## formatDecimal
+
+Represents a numeric value as String with exact number of fractional digits.
```
[the remark this answers is not in the document]
Comment (zvonand) [reply docs/en/sql-reference/functions/type-conversion-functions.md:742]: Why 'can'? the whole purpose of this is to specify this number, it is a required argument. We already have plain `toString` to do it w/o specifying the result scale

[the remark this answers is not in the document]
Comment (rschu1ze) [reply docs/en/sql-reference/functions/type-conversion-functions.md:742]: Okay. `Converts a numeric value to String with the number of fractional digits in the output specified by the user.`

**Review by rschu1ze — APPROVED** [review]

**Comment by zvonand** [comment]
Let's wait for tests to pass (I think I fixed the issues found by sanitizer)

**Comment by rschu1ze** [comment]
Alright ... I also pushed some more cosmetics (the main thing is that we no longer place a generic named class `Processor` into the `DB` namespace).

**Comment by zvonand** [comment]
@rschu1ze after I pulled your commit, it does not build, will take a look

**Pull request merged**